cranelift/codegen/src/machinst/vcode.rs

```diff
@@ -155,9 +155,17 @@ impl<I: VCodeInst> VCodeBuilder<I> {
 
     /// Set the type of a VReg.
     pub fn set_vreg_type(&mut self, vreg: VirtualReg, ty: Type) {
+        // TO_ASK .. Why is this code pushing new default values
```

Comment:
Otherwise `self.vcode.vreg_types[vreg.get_index()]` would panic on out of bounds indexing.

Reply:
Yes I see ... initially I wasn't sure why the need for the loop. I don't see a grow() for vectors but perhaps instead of looping we can replace with a resize statement? Something like:

> self.vcode.vreg_types.resize(ir::types::I8, vreg.get_index() - self.vcode.vreg_types.len())

Reply:
`vreg.get_index()` may be smaller than `self.vcode.vreg_types.len()`, in which case the substraction will underflow.

Reply:
I kept the guard. It's just local, but it looks like this:

>         if self.vcode.vreg_types.len() < vreg.get_index() {
>             self.vcode.vreg_types.resize(
>                 vreg.get_index() - self.vcode.vreg_types.len(),
>                 ir::types::I8,
>             )
>         }
>         self.vcode.vreg_types[vreg.get_index()] = ty;

as opposed to this with the while loop:

>         while self.vcode.vreg_types.len() <= vreg.get_index() {
>             self.vcode.vreg_types.push(ir::types::I8); // Default type.
>         }
>         self.vcode.vreg_types[vreg.get_index()] = ty;

I assumed that top "if-branch would be more efficient but not sure what rust is doing under the covers for the resize compared to the push.

Reply:
In general, a resize may reallocate all the memory you need at once, while several pushes in a row may cause several series of reallocation, so `resize` is to be preferred. (I don't know if rustc even gets to use some kind of SIMD tricks to fill the memory with the default value.)

Reply:
@Bnjbvr ... Cool thanks. Good to know.

Reply:
@bnjbvr .. Cool thanks.

Reply:
@bnjbvr Ok thanks. That SIMD question is an interesting one.